control_plane/attachment_service/src/service.rs

```diff
@@ -742,42 +745,275 @@ impl Service {
             (waiters, response_shards)
         };
 
-        let deadline = Instant::now().checked_add(Duration::from_secs(5)).unwrap();
+        self.await_waiters(waiters).await?;
+
+        Ok(TenantCreateResponse {
+            shards: response_shards,
+        })
+    }
+
+    /// Helper for functions that reconcile a number of shards, and would like to do a timeout-bounded
+    /// wait for reconciliation to complete before responding.
+    async fn await_waiters(
+        &self,
+        waiters: Vec<ReconcilerWaiter>,
+    ) -> Result<(), ReconcileWaitError> {
+        let deadline = Instant::now().checked_add(Duration::from_secs(30)).unwrap();
         for waiter in waiters {
             let timeout = deadline.duration_since(Instant::now());
             waiter.wait_timeout(timeout).await?;
         }
-        Ok(TenantCreateResponse {
-            shards: response_shards,
-        })
+
+        Ok(())
     }
 
-    pub(crate) async fn tenant_timeline_create(
+    /// This API is used by the cloud control plane to do coarse-grained control of tenants:
+    /// - Call with mode Attached* to upsert the tenant.
+    /// - Call with mode Detached to switch to PolicyMode::Detached
+    ///
+    /// In future, calling with mode Secondary may switch to a detach-lite mode in which a tenant only has
+    /// secondary locations.
+    pub(crate) async fn tenant_location_config(
```

Comment:
First, here's a suggestion to explain the danger around generation numbers.

```suggestion
    /// This API is used by the cloud control plane to migrate unsharded tenants that it created
    /// directly with pageservers into this service.
    /// 
    /// Cloud control plane MUST NOT continue issuing GENERATION NUMBERS for this tenant once it
    /// has attempted to call this API. Failure to oblige to this rule may lead to S3 corruption.
    /// Think of the first attempt to call this API as a transfer of absolute authority over the
    /// tenant's source of generation numbers.
    pub(crate) async fn tenant_location_config(
```

Second, I think it's a mistake to
1. overload this API with the "import unsharded tenants" functionality and
2. provide this as a virtual-pageserver-API at all.

Why? Because the semantics of import are simply not the same as the equally-named pageserver API.
I'm most worried about control plane bugs / seemlingly unrelated control plane code not properly keeping track of the special status.
For example, livemigration doesn't work once you import a tenant into this service:
Say you have a tenant that control plane migrates into this service.
Then someone clicks live migration on that tenant in control plane.
This API here will happily detach the tenant, and control plane will attach the tenant somwhere else, using its stale generation numbers count.

Reply:
In the near term, https://github.com/neondatabase/cloud/issues/9446 is the plan to enforce these rules in the control plane.  That should happen after we deploy in staging, but before we deploy in production.  I've added a comment there to highlight the correctness impact, and suggest that we could add a simpler/stricter enforcement by poisoning the generation state in the control plane, so that we don't rely on all relevant code locations to check if they're dealing with such a tenant: they'd just fail to acquire a generation number.

Once we have #9446, it opens the door to diverging the APIs a bit (e.g. to create a cleaner onboarding API, and call that conditionally from the control plane instead of this one).  That might be something we can do ourselves, as it would be quite a localized `if` statement.

Reply:
I'll pull the suggested comment change into https://github.com/neondatabase/neon/pull/6521